README.md
## Introduction
**hen** builds and packages your vala projects in a simple declarative way.

It implements sane defaults and uses [cmake](http://cmake.org/) under the hood, meaning it works where cmake works (nearly everywhere)

All is needed in **only one** `project.hen` file describing your whole project.

```java
# This new application will conquer the world
application (
    AUTHOR
        "Your self <your.self@home.com>"
    HOMEPAGE
        "https://home.com/your-application"      
    LICENSE
        "Apache 2.0"  
    NAME
        your-application
    TITLE
        "This application will be great"
    VERSION
        "1.0"
    SOURCE_PATH
        src
    PACKAGES
        gtk+-3.0
)

```
Then type `hen build` to build your application and `hen package` to create `.deb` files ready for distribution.

![sample](docs/hen-screenshot.png) 

If you want to run your application, type `hen run` or `hen debug` to start a `gdb` debugging session, or any of the other [hen commands](docs/commands.md)

> note: the `project.hen` is a cmake file you can use whatever cmake instructions to spice it.

## Features: 
  - build and install different [binary prototypes](docs/prototypes.md) gui/console applications, libraries and plugins using a simple declarative syntax
  - generate a wide array of files for you: .desktop and icons for applications, .pc .deps for lbraries
  - one stop commmand script `./hen` for all the common tasks: build, rebuild, install, etc.
  - auto-install itself and more important *auto-update* via `./hen update`

## Getting started 

```
# install the script locally 
curl -sL  http://i-hate-farms.github.io/hen/install | bash -  
# create a sample project
./hen init
```

More information in the [gettting started section](docs/getting-started.md).

## Rodmap to 1.0 
  - [ ] handle application icons and images 
  - [ ] handle po files and translations
  - [ ] finish up valadoc task
  - [ ] documentation and test suite

## Samples

You can find samples for: 
  - [vala-stacktrace][1]: a library and a sample command line test program
  - [eidete][2]: an elementary gtk app
  - [useraccounts][4]: an elementary [switchboard][3] plug 
  - [webcontracts][5]: an elementary plug shipping with a cli application 
  - [gtk-test][6]: a simple Gtk application
  - [cli-test][7]: a simple command line application

[1]: https://github.com/PerfectCarl/vala-stacktrace
[2]: https://code.launchpad.net/~name-is-carl/eidete/use-elementary.cmake
[3]: https://launchpad.net/switchboard
[4]: https://code.launchpad.net/~name-is-carl/switchboard-plug-useraccounts/use-elementary.cmake
[5]: https://code.launchpad.net/~elementary-apps/webcontracts/fix-for-freya
[6]: none
[7]: none


## [Documentation](docs/doc.md) 

## [Changelog](CHANGELOG.md)

## License
Released under Apache 2.0 license


CHANGELOG.md
- support uninstall
- support system icons for ICON
- support po translation files
- support threads

* 0.9.27
  - fix install output
  - fix packaging output
  - fix weird libivy package error

* 0.9.25
  - run, debug and package depends on build task
  - fix libraries deployed partly in /usr instead of /usr/bin
  - add ivy to deps.list
  - add spores to hen setup process
  - use APPEND_STRING instead of APPEND

* 0.9.24
  - fix messages
  - remove valadoc
  - fix muti-targets (a deal is a deal)
  - add C_OPTIONS to deal with gcc -lm

* 0.9.23
  - fix bugs in multi target
  - enable SRC to be empty
  - tweak messages
  - fixing regression: library not installed in /usr/lib because generated CMakeLists.txt date was off

* 0.9.22
  - add libvala-0.28
  - fix ./hen build bugs

* 0.9.21
  - fails if there is nothing to build
  - better messages
  - display the list of installed dependencies
  - FORMAT: BINARY_NAME => NAME

* 0.9.20
 - new valadoc style CSS
 - fix install of libraries in /usr and not /usr/local/lib as the path
   is not used by ldconfig by default
 - add package task to build debian files
 - FORMAT: add AUTHOR, COPYRIGHT, LICENSE and HOMEPAGE
 - add run and debug tasks
 - add init task
 - FORMAT: introduce the project.hen file
 * add valadoc task

* 0.9.13
 - add force-update task
 - fix regression if your project requires built-in dependencies (eidete)
 - fix regression in installing desktop icon (eidete-videobin)

* 0.9.12
 - rename cmake folder to .hen
 - add tasks: valadoc
 - first stab at deb building and providing default icons

* 0.9.11
 - self install
 - add install task
 - hen auto-install its dependencies (valac, cmake, build-essential) the first time

* 0.9.10
 - generate desktop and icon file
 - add clean task
 - improve onboarding experience
 - more documentation

* 0.9.9
 - only install apt packages if not present in the system (avoid asking for sudo without reason)
 - fix issues with gtk3.0 and ox
 - refactoring
 - version number can be added to packages (they are ignored for now)

* 0.9.8
 - misc fixes for apt dependencies
 - misc formatting fixes
 - hen is now updated with updates

* 0.9.7
 - add an internal release task

* 0.9.6
 - apt packages are installed automatically

* 0.9.5
 - fix comments
 - add rebuild task

* 0.9.4
 - add list options
 - add colors
 - fix library task (vala-stacktrace working)
 - add --target-glib 2.32 by default (and emit a warning if set in the cmake file)

* 0.9.3
 - refactoring (eidete working)
 - add build and prepare options

* 0.9.1:
 - add `hen.sh`
 - publish 0.9.1 to github

* 0.9:
- 0.1 : initial release
- 0.2 : support schemas
- 0.3 : support app
- 0.4 : misc fixes
- 0.5 : support shared libraries
- 0.6 : support cli apps. Support C_OPTIONS
- 0.7 : add full dependencies support
- 0.8 : fix hard coded vapi and add local ./vapi folder
- 0.9 : support VALA_DEFINES
- 0.10: add RELEASE_NAME
- 0.11: adding support for desktop and icon files
- 0.12: setting a default for CMAKE_INSTALL_PREFIX if left empty
- 0.13: use BUILD_TYPE for the build and set vala define BUILD_IS_DEBUG and BUILD_IS_RELEASE
- 0.14: support and install contracts
- 0.15: implement *.vala and *.c files


hen.sh
#!/bin/bash
# File history 
#  - 0.1: update and help
#  - 0.2: build and prepare
#  - 0.3: list

OPERATION="build"

DEST_FOLDER=./cmake
TEMP_FOLDER=./hen-tmp
SOURCE_URL=https://github.com/I-hate-zoos/hen/raw/master/dist/
PACKAGE_NAME=hen-latest.zip
PACKAGE_URL=${SOURCE_URL}/${PACKAGE_NAME}
VERSION_URL=${SOURCE_URL}/Hen-VERSION.txt

PACKAGE_FILE=${TEMP_FOLDER}/${PACKAGE_NAME}
VERSION_FILE=${TEMP_FOLDER}/Hen-VERSION.txt

BUILD_FOLDER=./build

red='\033[0;31m'
white='\033[1;37m'
NC='\033[0m' # No Color

if [ -n "$1" ]; then
    OPERATION=$1
fi

do_update () {
    if [ -d ${TEMP_FOLDER} ] ; then
    	rm -rf ${TEMP_FOLDER}
    fi
    mkdir ${TEMP_FOLDER} 
    echo "Checking for newer version..."
    wget --quiet -P ${TEMP_FOLDER} ${VERSION_URL}
    SERVER_VERSION=`cat ${VERSION_FILE}`
    if [ -d ${DEST_FOLDER} ] ; then
    	LOCAL_VERSION=`cat ${DEST_FOLDER}/Hen-VERSION.txt`
    	echo " Local version : ${LOCAL_VERSION}"
    fi 

    echo -e " Server version: ${white}${SERVER_VERSION}${NC}"

    # Is an update required?
	if [ "${SERVER_VERSION}" == "${LOCAL_VERSION}" ]; then
    	echo "Your version is up to date (${LOCAL_VERSION})"
	else
		echo "Newer version found!"
		echo "Getting the new files from server..."
	    wget --quiet -P ${TEMP_FOLDER} ${PACKAGE_URL}
	    if [ -d ${DEST_FOLDER} ] ; then
	    	rm -rf ${DEST_FOLDER}
	    fi 
	    unzip -q ${PACKAGE_FILE} -d ${DEST_FOLDER}

	    # Cleaning up
	    rm -rf ${TEMP_FOLDER}

	    echo -e "${white}Hen updated to ${SERVER_VERSION}${NC}"
    fi

}

do_prepare () {
      # if the Hen version can't be found, it means that the 
      # Hen needs to be installed
      if [ ! -f "${DEST_FOLDER}/Hen-VERSION.txt" ] ; then
        do_update 
      fi 
      echo -e "${white}Preparing your build ...${NC}"
      if [ -d ${BUILD_FOLDER} ] ; then
        rm -rf ${BUILD_FOLDER}
      fi 
      mkdir ${BUILD_FOLDER}
      cd ${BUILD_FOLDER}
      cmake .. 
}

do_build () {
      if [ ! -d ${BUILD_FOLDER} ] ; then
        do_prepare 
      fi 
      echo -e "${white}Building your application ...${NC}"
      cd ${BUILD_FOLDER}
      make 
      cd ..        
}

do_help () {
	if [ -d ${DEST_FOLDER} ] ; then
    	LOCAL_VERSION=`cat ${DEST_FOLDER}/Hen-VERSION.txt`
    	echo -e "${white}Hen ${LOCAL_VERSION}${NC}"
    	echo ""
      cat ./cmake/Hen-HELP.txt
  else
    echo -e "${white}No help is available${NC}. Please update hen with: $0 update"
  fi 
}

do_list () {
  echo -e "${white}Hen${NC} defines the following make tasks"
  echo -e "  - ${white}[default]${NC} : build your application along with extra files if applicable (desktop, vapi, pc, deps...)"
  echo -e "  - ${white}install${NC} : install your application to the previously defined PREFIX (./usr by default)"
  echo -e "  - ${white}pot${NC} : generates the translation files in the po folder"
  echo "" 
  echo "You can invoke those tasks in the ./build folder: "
  echo "  cd build"
  echo "  make pot"
}


case "$OPERATION" in
   "update") do_update  
   ;;
   "prepare") do_prepare
   ;;
   "build")do_build
   ;;   
   "list")do_list
   ;;
   "help")do_help
   ;;
   "--help")do_help
   ;;     
   "--?")do_help
   ;;   
   "/?")do_help
   ;;  
   "-h")do_help
   ;;     
esac